### PR TITLE
(2728) Allow opposite ODA type partner countries

### PR DESCRIPTION
**Pull request by yndajas**
## Changes in this PR

This splits the ISPF partner countries field into two ODA type-specific fields: ISPF ODA partner countries and ISPF non-ODA partner countries. This is to allow non-ODA partner countries on ODA activities. The UI form, activity show page, CSV export and CSV import are all updated to reflect this change

A data migration is included for shifting staging and training data over to the new columns, after which we can create another migration for removing the old column

#### Show page

Screenshot 2023-01-17 at 18 51 36

#### Form

image

#### Show page (ODA)

Screenshot 2023-01-17 at 18 50 06

#### Form (ODA)

image

image

## Next steps

- [ ] Is an ADR required? An ADR should be added if this PR introduces a change to the architecture.
- [x] Is a changelog entry required? An entry should always be made in `CHANGELOG.md`, unless this PR is a small tweak which has no impact outside the development team.
- [ ] Do any environment variables need amending or adding?
- [ ] Have any changes to the XML been checked with the IATI validator? See [XML Validation](https://github.com/UKGovernmentBEIS/beis-report-official-development-assistance/blob/develop/doc/xml-validation.md)

### CODE REVIEW

**Review by Gweaton — COMMENTED** [review]
Looking good so far - a couple of things we've discussed.

File: app/helpers/codelist_helper.rb [hunk 1]
```diff
@@ -139,8 +139,15 @@ def ispf_themes_options
     Codelist.new(type: "ispf_themes", source: "beis").to_objects_with_description
   end
 
-  def ispf_partner_country_options(is_oda:)
-    Codelist.new(type: "ispf_partner_countries", source: "beis").to_partner_country_options(is_oda: is_oda)
+  def ispf_partner_country_options(oda:, none: true)
```
[the remark this answers is not in the document]
Comment (yndajas) [reply app/helpers/codelist_helper.rb:142]: I thought that myself when writing it, but was being lazy!

File: app/models/activity.rb [hunk 2]
```diff
@@ -682,8 +693,12 @@ def is_non_oda_project?
   end
 
   def ispf_partner_countries_none_is_exclusive
-    if ispf_partner_countries&.include?("NONE") && ispf_partner_countries.size > 1
-      errors.add(:ispf_partner_countries, I18n.t("activerecord.errors.models.activity.attributes.ispf_partner_countries.none_exclusive"))
+    type = is_oda ? "oda" : "non_oda"
+    field = :"ispf_#{type}_partner_countries"
+    value = send(field)
```
[the remark this answers is not in the document]
Comment (yndajas) [reply app/models/activity.rb:698]: Instead of `value`?

File: app/models/ispf_non_oda_partner_country.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class IspfNonOdaPartnerCountry < IspfPartnerCountry
```
Comment (yndajas) [inline app/models/ispf_non_oda_partner_country.rb:1]: Use "partner countries" in commit title, in line with other commits

**Review by Gweaton — APPROVED** [review]

**Pull request merged**